File: python/coffee_shop.py
# Instances of CoffeeFlavour will be the Flyweights
class CoffeeFlavour:

    def __init__(self, newFlavor):
        self.name = newFlavor

# Menu acts as a factory and cache for CoffeeFlavour flyweight objects
class Menu:

    def __init__(self):
        self.flavours = {}

    def lookup(self, flavourName):
        if flavourName not in self.flavours:
            self.flavours[flavourName] = CoffeeFlavour(flavourName)
        return self.flavours[flavourName]

class Order:

    def __init__(self, tableNumber, flavor):
        self.tableNumber = tableNumber
        self.flavour = flavor

    def serve(self):
        pass

class CoffeeShop:

    def __init__(self):
        self.orders = []
        self.menu = Menu()

    def takeOrder(self, flavourName, table):
        flavour = self.menu.lookup(flavourName)
        order = Order(table, flavour)
        self.orders.append(order)

    def service(self):
        for order in self.orders:
            order.serve()

File: java/CoffeeShop.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// Instances of CoffeeFlavour will be the Flyweights
class CoffeeFlavour {
	private final String name;

	CoffeeFlavour(String newFlavor) {
		this.name = newFlavor;
	}

	@Override
	public String toString() {
		return name;
	}
}

// Menu acts as a factory and cache for CoffeeFlavour flyweight objects
class Menu {
	private Map<String, CoffeeFlavour> flavours = new HashMap<String, CoffeeFlavour>();

	CoffeeFlavour lookup(String flavorName) {
		if (!flavours.containsKey(flavorName))
			flavours.put(flavorName, new CoffeeFlavour(flavorName));
		return flavours.get(flavorName);
	}

	int totalCoffeeFlavoursMade() {
		return flavours.size();
	}
}

class Order {
	private final int tableNumber;
	private final CoffeeFlavour flavour;

	Order(int tableNumber, CoffeeFlavour flavor) {
		this.tableNumber = tableNumber;
		this.flavour = flavor;
	}

	void serve() {
	}
}

class CoffeeShop {
	private final List<Order> orders = new ArrayList<Order>();
	private final Menu menu = new Menu();

	void takeOrder(String flavourName, int table) {
		CoffeeFlavour flavour = menu.lookup(flavourName);
		Order order = new Order(table, flavour);
		orders.add(order);
	}

	void service() {
		for (Order order : orders)
			order.serve();
	}

}
